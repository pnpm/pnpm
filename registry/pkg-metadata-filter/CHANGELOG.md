# @pnpm/registry.pkg-metadata-filter

## 1000.0.0

### Major Changes

- 4a2d871: Initial release.

### Patch Changes

- Updated dependencies [4a2d871]
  - @pnpm/registry.types@1000.0.0
