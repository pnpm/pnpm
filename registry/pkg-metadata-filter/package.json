{
  "name": "@pnpm/registry.pkg-metadata-filter",
  "version": "1000.1.0",
  "description": "Filters the package metadata from the registry",
  "keywords": [
    "pnpm",
    "pnpm10",
    "npm",
    "pkg-doc"
  ],
  "license": "MIT",
  "funding": "https://opencollective.com/pnpm",
  "repository": "https://github.com/pnpm/pnpm/tree/main/registry/pkg-metadata-filter",
  "homepage": "https://github.com/pnpm/pnpm/tree/main/registry/pkg-metadata-filter#readme",
  "bugs": {
    "url": "https://github.com/pnpm/pnpm/issues"
  },
  "type": "commonjs",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "exports": {
    ".": "./lib/index.js"
  },
  "files": [
    "lib",
    "!*.map"
  ],
  "scripts": {
    "lint": "eslint \"src/**/*.ts\" \"test/**/*.ts\"",
    "_test": "jest",
    "test": "pnpm run compile && pnpm run _test",
    "prepublishOnly": "pnpm run compile",
    "compile": "tsc --build && pnpm run lint --fix"
  },
  "dependencies": {
    "@pnpm/registry.types": "workspace:*",
    "semver": "catalog:"
  },
  "peerDependencies": {
    "@pnpm/logger": "catalog:"
  },
  "devDependencies": {
    "@pnpm/logger": "workspace:*",
    "@pnpm/registry.pkg-metadata-filter": "workspace:*",
    "@types/semver": "catalog:"
  },
  "engines": {
    "node": ">=18.12"
  },
  "jest": {
    "preset": "@pnpm/jest-config"
  }
}
