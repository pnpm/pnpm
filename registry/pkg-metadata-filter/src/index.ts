import { globalWarn } from '@pnpm/logger'
import { type PackageMetadataWithTime } from '@pnpm/registry.types'
import semver from 'semver'

export function filterPkgMetadataByPublishDate (pkgDoc: PackageMetadataWithTime, publishedBy: Date): PackageMetadataWithTime {
  const versionsWithinDate: PackageMetadataWithTime['versions'] = {}
  for (const version in pkgDoc.versions) {
    if (!Object.hasOwn(pkgDoc.versions, version)) continue
    const timeStr = pkgDoc.time[version]
    if (timeStr && new Date(timeStr) <= publishedBy) {
      versionsWithinDate[version] = pkgDoc.versions[version]
    }
  }

  const distTagsWithinDate: PackageMetadataWithTime['dist-tags'] = {}
  const allDistTags = pkgDoc['dist-tags'] ?? {}
  const parsedSemverCache = new Map<string, semver.SemVer>()
  function tryParseSemver (semverStr: string): semver.SemVer | null {
    let parsedSemver = parsedSemverCache.get(semverStr)
    if (!parsedSemver) {
      try {
        parsedSemver = new semver.SemVer(semverStr, true)
      } catch {
        return null
      }
      parsedSemverCache.set(semverStr, parsedSemver)
    }
    return parsedSemver
  }
  for (const tag in allDistTags) {
    if (!Object.hasOwn(allDistTags, tag)) continue
    const distTagVersion = allDistTags[tag]
    if (versionsWithinDate[distTagVersion]) {
      distTagsWithinDate[tag] = distTagVersion
      continue
    }
    // Repopulate the tag to the highest version available within date that has the same major as the original tag's version
    const originalSemVer = tryParseSemver(distTagVersion)
    if (!originalSemVer) continue
    const originalIsPrerelease = (originalSemVer.prerelease.length > 0)
    let bestVersion: string | undefined
    for (const candidate in versionsWithinDate) {
      if (!Object.hasOwn(versionsWithinDate, candidate)) continue
      const candidateParsed = tryParseSemver(candidate)
      if (
        !candidateParsed ||
        candidateParsed.major !== originalSemVer.major ||
        (candidateParsed.prerelease.length > 0) !== originalIsPrerelease
      ) continue
      if (!bestVersion) {
        bestVersion = candidate
      } else {
        try {
          const candidateIsDeprecated = pkgDoc.versions[candidate].deprecated != null
          const bestVersionIsDeprecated = pkgDoc.versions[bestVersion].deprecated != null
          if (
            (semver.gt(candidate, bestVersion, true) && (bestVersionIsDeprecated === candidateIsDeprecated)) ||
            (bestVersionIsDeprecated && !candidateIsDeprecated)
          ) {
            bestVersion = candidate
          }
        } catch (err) {
          globalWarn(`Failed to compare semver versions ${candidate} and ${bestVersion} from packument of ${pkgDoc.name}, skipping candidate version.`)
        }
      }
    }
    if (bestVersion) {
      distTagsWithinDate[tag] = bestVersion
    }
  }

  return {
    ...pkgDoc,
    versions: versionsWithinDate,
    'dist-tags': distTagsWithinDate,
  }
}
