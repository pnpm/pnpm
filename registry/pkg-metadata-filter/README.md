# @pnpm/registry.pkg-metadata-filter

> Filters the package metadata from the registry

<!--@shields('npm')-->
[![npm version](https://img.shields.io/npm/v/@pnpm/registry.pkg-metadata-filter.svg)](https://www.npmjs.com/package/@pnpm/registry.pkg-metadata-filter)
<!--/@-->

## Installation

```
pnpm add @pnpm/registry.pkg-metadata-filter
```

## Usage

```ts
import { filterPkgMetadataByPublishDate } from '@pnpm/registry.pkg-metadata-filter'

const pkgDoc = await (await fetch('https://registry.npmjs.org/is-odd')).json()

// Keep only those versions in the document that were published before Jan 1 2023.
filterPkgMetadataByPublishDate(pkgDoc, new Date('2023-01-01'))
```

## License

MIT
