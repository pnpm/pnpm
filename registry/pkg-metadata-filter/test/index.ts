import { filterPkgMetadataByPublishDate } from '@pnpm/registry.pkg-metadata-filter'

test('filterPkgMetadataByPublishDate', () => {
  const cutoff = new Date('2020-04-01T00:00:00.000Z')
  const name = 'dist-tag-date'
  expect(filterPkgMetadataByPublishDate({
    name,
    versions: {
      '3.0.0': {
        name,
        version: '3.0.0',
        dist: { tarball: `https://registry.npmjs.org/${name}/-/${name}-3.0.0.tgz`, shasum: '' },
      },
      '3.1.0': {
        name,
        version: '3.1.0',
        dist: { tarball: `https://registry.npmjs.org/${name}/-/${name}-3.1.0.tgz`, shasum: '' },
        deprecated: 'This version is deprecated',
      },
      '3.2.0': {
        name,
        version: '3.2.0',
        dist: { tarball: `https://registry.npmjs.org/${name}/-/${name}-3.2.0.tgz`, shasum: '' },
      },
      '2.9.9': {
        name,
        version: '2.9.9',
        dist: { tarball: `https://registry.npmjs.org/${name}/-/${name}-2.9.9.tgz`, shasum: '' },
      },
    },
    'dist-tags': {
      latest: '3.2.0',
    },
    time: {
      '2.9.9': '2020-01-01T00:00:00.000Z',
      '3.0.0': '2020-02-01T00:00:00.000Z',
      '3.1.0': '2020-03-01T00:00:00.000Z',
      '3.2.0': '2020-05-01T00:00:00.000Z',
    },
  }, cutoff)).toMatchSnapshot()
})
