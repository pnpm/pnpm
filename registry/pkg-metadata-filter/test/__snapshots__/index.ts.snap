// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterPkgMetadataByPublishDate 1`] = `
{
  "dist-tags": {
    "latest": "3.0.0",
  },
  "name": "dist-tag-date",
  "time": {
    "2.9.9": "2020-01-01T00:00:00.000Z",
    "3.0.0": "2020-02-01T00:00:00.000Z",
    "3.1.0": "2020-03-01T00:00:00.000Z",
    "3.2.0": "2020-05-01T00:00:00.000Z",
  },
  "versions": {
    "2.9.9": {
      "dist": {
        "shasum": "",
        "tarball": "https://registry.npmjs.org/dist-tag-date/-/dist-tag-date-2.9.9.tgz",
      },
      "name": "dist-tag-date",
      "version": "2.9.9",
    },
    "3.0.0": {
      "dist": {
        "shasum": "",
        "tarball": "https://registry.npmjs.org/dist-tag-date/-/dist-tag-date-3.0.0.tgz",
      },
      "name": "dist-tag-date",
      "version": "3.0.0",
    },
    "3.1.0": {
      "deprecated": "This version is deprecated",
      "dist": {
        "shasum": "",
        "tarball": "https://registry.npmjs.org/dist-tag-date/-/dist-tag-date-3.1.0.tgz",
      },
      "name": "dist-tag-date",
      "version": "3.1.0",
    },
  },
}
`;
